text/0003-webcontents-invoke.md

```diff
@@ -0,0 +1,373 @@
+
+2. Frame Navigation: If the target frame navigates before responding to the IPC message.
+
+   a. cross-site navigation: If a render frame navigates to a different site (not just a different page), it will be swapped out, and a new render frame host will be created (process isolation).
+
+   b. same-site navigation: If the frame navigates to a different page within the same site, the same render frame host will be used.
+
+
+      > Current Solution: Whether same or cross site navigation, the ipcRenderer will respond with a resolved value from the corresponding WebFrameMain/RenderFrameHost that made the call. Ideally we want to reject the response if it is not the same page we intended to invoke from the main process, but there is no way to control that same site navigation re-uses the same frame with the same context. 
+
+      >The target frame is decided when the main process makes the invoke because the WebFrameMain/RenderFrameHost corresponds to the ElectronRenderer/RendererAPI. So if the render frame navigates away after invoking the request, the promise will be resolved with the handler from the WebFrameMain/RenderFrameHost that made the call. 
+      
+     >The implementation checks that the routing ID of the main frame (RenderFrameHost) matches the routing ID of the web frame that received the invoke call. This ensures that the main process receives a response from the intended web frame and not a different one. However, this check may be redundant because the ElectronRenderer/RendererAPI always corresponds to the calling WebFrameMain/RenderFrameHost, so the routing IDs should never be different.
+
+3. No Registered Handler: If the renderer process does not have a registered handler for the specified channel, the promise should be rejected.
+
+   > Current Solution: When page changes, whether same or cross site navigation, the renderer process updates and new handlers are registered specific to that renderer process. So if a call from main is made before a handler registers or to a handler that does not exist, the promise will be rejected.
+
+### Edge Cases Not Addressed
+
+4. Multiple Frames: The `webContents.invoke()` method supports invoking from the mainFrame WebFrameMain. Future area of development is being able to invoke from a specific RenderFrameHost/ WebFrameMain instance by providing the routingId/ frameId option. This allows for communication with individual frames within a single webContents instance.
+
+
+## Drawbacks
+
+- Introducing new APIs may increase the complexity of the Electron IPC and documentation. This can potentially introduce new bugs or edge cases for the team to support.
+- Implementing timeout functionality and handling edge cases may introduce additional overhead and potential performance implications.
+- The timeout mechanism may not be suitable for all use cases, as some operations in the renderer process may take longer than the configured timeout period.
```

Comment:
Yeah agreed! Like you mentioned, the implementation doesn't differ from what we already have in `ipcRenderer.invoke` so it's not additional work to maintain. For developers, we can always add warnings in the documentation guide when I work on adding a new section in our "Inter-Process Communication" so they are aware of edge cases.

Reply:
Following the "unresponsive" pattern for renderers may be the right move here. Not sure how this could fit into the spec but I'm imagining we may want to convey certain information about renderer state that Electron can infer that developers may struggle to identify.

E.g. Rejecting a promise with specific error codes if:
* The renderer is gone (crashed or other)
* The renderer has navigated since the request was dispatched
* The renderer has gone unresponsive while we were waiting for the response

Having some kind of `retry()` thing may help.

Might have missed it in the spec but I'm wondering what the exact semantics we're going for around sending a request to a booting renderer RE queuing.

E.g. 
```
* boot renderer
* send request "give data"
* renderer receives request as electron is alive but has no listener yet
  * request is failed due to no listener
* renderer runs preload
* renderer loads and adds listener
  * Maybe we should have queued the request till the handler was added? Indefinite timeout until the renderer is booted? Declare loaded listeners in the main process and allow folks to say:

webContents.ipc.onHandler('foo', () => {
  webContents.ipc.invoke('foo');
});

await webContents.ipc.waitForHandler('foo');
```

Thinking of common patterns around booting and waiting for data and how to handle those races via the API instead of letting folks build a "ready" signal themselves.

Reply:
Instead of callback and promise APIs, you could have one API that differs by arguments:

```
// callback style
webContents.ipc.on('foo', handleFoo);
// promise style
await webContents.ipc.on('foo');
// or only promise for once?
await webContents.ipc.once('foo');
```